pkg/controller/petset/identity_mappers.go

```diff
@@ -0,0 +1,202 @@
+	api_pod "k8s.io/kubernetes/pkg/api/pod"
+	"k8s.io/kubernetes/pkg/apis/apps"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+const svcLocalSuffix = "svc.cluster.local"
+
+// identityMapper is an interface for assigning identities to a pet.
+// All existing identity mappers just append "-(index)" to the petset name to
+// generate a unique identity. This is used in claims/DNS/hostname/petname
+// etc. There's a more elegant way to achieve this mapping, but we're
+// taking the simplest route till we have data on weather users will need
+// more customization.
+type identityMapper interface {
+	AddIdentity(id string, pet *api.Pod)
+	GetIdentity(pod *api.Pod) string
+}
+
+func NewIdentityMappers(ps *apps.PetSet) []identityMapper {
+	return []identityMapper{
+		&NameIdentityMapper{ps},
+		&NetworkIdentityMapper{ps},
+		&VolumeIdentityMapper{ps},
+	}
+}
+
+type NetworkIdentityMapper struct {
+	ps *apps.PetSet
+}
+
+func (n *NetworkIdentityMapper) AddIdentity(id string, pet *api.Pod) {
+	pet.Annotations[api_pod.PodHostnameAnnotation] = fmt.Sprintf("%v-%v", n.ps.Name, id)
+	pet.Annotations[api_pod.PodSubdomainAnnotation] = n.ps.Spec.ServiceName
+	return
+}
+
+func (n *NetworkIdentityMapper) GetIdentity(pet *api.Pod) string {
+	return n.String(pet)
+}
+
+func (n *NetworkIdentityMapper) String(pet *api.Pod) string {
+	hostname := pet.Annotations[api_pod.PodHostnameAnnotation]
+	subdomain := pet.Annotations[api_pod.PodSubdomainAnnotation]
+	return fmt.Sprintf("%v.%v.%v.%v", hostname, subdomain, n.ps.Namespace, svcLocalSuffix)
+}
+
+type VolumeIdentityMapper struct {
+	ps *apps.PetSet
+}
+
+func (v *VolumeIdentityMapper) AddIdentity(id string, pet *api.Pod) {
+	petVolumes := []api.Volume{}
+	petClaims := GetClaims(id, v.ps)
+
+	// These volumes will all go down with the pod. If a name matches one of
+	// the claims in the pet set, it gets clobbered.
+	podVolumes := map[string]api.Volume{}
+	for _, podVol := range pet.Spec.Volumes {
+		podVolumes[podVol.Name] = podVol
+	}
+
+	// Insert claims for the idempotent petSet volumes
+	for name, claim := range petClaims {
+		// Volumes on a pet for which there are no associated claims on the
+		// petset are pod local, and die with the pod.
+		podVol, ok := podVolumes[name]
+		if ok {
+			glog.V(4).Infof("Overwriting existing volume source %v", podVol.Name)
+		}
+		newVol := api.Volume{
+			Name: name,
+			VolumeSource: api.VolumeSource{
+				PersistentVolumeClaim: &api.PersistentVolumeClaimVolumeSource{
+					ClaimName: claim.Name,
+					ReadOnly:  false,
+				},
+			},
+		}
+		petVolumes = append(petVolumes, newVol)
+	}
+
+	// Transfer any ephemeral pod volumes
+	for name, vol := range podVolumes {
+		if _, ok := petClaims[name]; !ok {
+			petVolumes = append(petVolumes, vol)
+		}
+	}
+	pet.Spec.Volumes = petVolumes
+	return
+}
+
+func (v *VolumeIdentityMapper) GetIdentity(pet *api.Pod) string {
+	// TODO: Make this a hash?
+	return v.String(pet)
+}
+
+func (v *VolumeIdentityMapper) String(pet *api.Pod) string {
```

Comment:
That's fine, just wasn't obvious from inspection why this way was cleaner.
Solvable with Godoc.

On Thu, Apr 28, 2016 at 10:55 PM, Prashanth B <notifications@github.com>
wrote:

> In pkg/controller/petset/identity_mappers.go
> <https://github.com/kubernetes/kubernetes/pull/24912#discussion_r61530451>
> :
>
> > +	// Transfer any ephemeral pod volumes
> > +	for name, vol := range podVolumes {
> > +		if _, ok := petClaims[name]; !ok {
> > +			petVolumes = append(petVolumes, vol)
> > +		}
> > +	}
> > +	pet.Spec.Volumes = petVolumes
> > +	return
> > +}
> > +
> > +func (v *VolumeIdentityMapper) GetIdentity(pet *api.Pod) string {
> > +	// TODO: Make this a hash?
> > +	return v.String(pet)
> > +}
> > +
> > +func (v *VolumeIdentityMapper) String(pet *api.Pod) string {
>
> It's like a custom string method for just the volume identity of the given
> pod. Without looking at ns/name you should be able to concat the String
> methods of all identity mappers and get something unique for this pet. If 2
> pets are somehow using the same pvc, they'll get the same volume id. Simply
> stringifying it is the simples way to produce a id hash (suggestions?).
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24912/files/0ecd23c7401b6f2bc0fb35c0101aa8d4d139511f..dfb99f88a2bbf772b26f7e31c4ce4878a04fb937#r61530451>
>